scanner/src/main/kotlin/ScanResultsStorage.kt

```diff
@@ -154,27 +154,30 @@ abstract class ScanResultsStorage : PackageBasedScanStorage {
             if (results.isEmpty()) {
                 results
             } else {
-                val scanResults = results.toMutableList()
+                val (matchingProvenance, nonMatchingProvenance) = results.partition { it.provenance.matches(pkg) }
 
-                // Only keep scan results whose provenance information matches the package information.
-                scanResults.retainAll { it.provenance.matches(pkg) }
-                if (scanResults.isEmpty()) {
+                if (matchingProvenance.isEmpty()) {
                     logger.debug {
                         "No stored scan results found for $pkg. The following entries with non-matching provenance " +
-                            "have been ignored: ${scanResults.map { it.provenance }}"
+                            "have been ignored: ${nonMatchingProvenance.map { it.provenance }}"
                     }
+
+                    matchingProvenance
```

Comment:
Can the code now return result entries which do match provenance, but do not match criteria in line 165?

Reply:
No, as this is inside the `matchingProvenance.isEmpty()` block.